runtime/domain.c

```diff
@@ -16,79 +18,1442 @@
+  }
+  CAMLassert(i < stw_domains.participating_domains);
+
+  /* swap passed domain to first free domain */
+  stw_domains.participating_domains--;
+  stw_domains.domains[i] =
+      stw_domains.domains[stw_domains.participating_domains];
+  stw_domains.domains[stw_domains.participating_domains] = dom;
+}
+
+static dom_internal* next_free_domain() {
+  if (stw_domains.participating_domains == Max_domains)
+    return NULL;
+
+  CAMLassert(stw_domains.participating_domains < Max_domains);
+  return stw_domains.domains[stw_domains.participating_domains];
+}
+
+#ifdef __APPLE__
+/* OSX has issues with dynamic loading + exported TLS.
+    This is slower but works */
+CAMLexport pthread_key_t caml_domain_state_key;
+static pthread_once_t key_once = PTHREAD_ONCE_INIT;
+
+static void caml_make_domain_state_key (void)
 {
-  if (Caml_state != NULL)
-    return;
+  (void) pthread_key_create (&caml_domain_state_key, NULL);
+}
+
+void caml_init_domain_state_key (void)
+{
+  pthread_once(&key_once, caml_make_domain_state_key);
+}
+
+#else
+CAMLexport __thread caml_domain_state* Caml_state;
+#endif
+
+/* Interrupt functions */
+static const uintnat INTERRUPT_MAGIC = (uintnat)(-1);
+
+Caml_inline void interrupt_domain(struct interruptor* s) {
+  atomic_store_rel(s->interrupt_word, INTERRUPT_MAGIC);
+}
+
+int caml_incoming_interrupts_queued(void)
+{
+  return atomic_load_acq(&domain_self->interruptor.interrupt_pending);
+}
+
+/* must NOT be called with s->lock held */
+static void stw_handler(caml_domain_state* domain);
+static uintnat handle_incoming(struct interruptor* s)
+{
+  uintnat handled = atomic_load_acq(&s->interrupt_pending);
+  CAMLassert (s->running);
+  if (handled) {
+    atomic_store_rel(&s->interrupt_pending, 0);
+
+    stw_handler(domain_self->state);
+  }
+  return handled;
+}
+
+static void handle_incoming_otherwise_relax (struct interruptor* self)
+{
+  if (!handle_incoming(self))
+    cpu_relax();
+}
+
+void caml_handle_incoming_interrupts(void)
+{
+  handle_incoming(&domain_self->interruptor);
+}
+
+int caml_send_interrupt(struct interruptor* target)
+{
+  /* signal that there is an interrupt pending */
+  CAMLassert(!atomic_load_acq(&target->interrupt_pending));
+  atomic_store_rel(&target->interrupt_pending, 1);
+
+  /* Signal the condition variable, in case the target is
+     itself waiting for an interrupt to be processed elsewhere */
+  caml_plat_lock(&target->lock);
+  caml_plat_broadcast(&target->cond); // OPT before/after unlock? elide?
+  caml_plat_unlock(&target->lock);
+
+  interrupt_domain(target);
+
+  return 1;
+}
+
+static void caml_wait_interrupt_serviced(struct interruptor* target)
+{
+  int i;
+
+  /* Often, interrupt handlers are fast, so spin for a bit before waiting */
+  for (i=0; i<1000; i++) {
+    if (!atomic_load_acq(&target->interrupt_pending)) {
+      return;
+    }
+    cpu_relax();
+  }
+
+  {
+    SPIN_WAIT {
+      if (!atomic_load_acq(&target->interrupt_pending))
+        return;
+      cpu_relax();
```

Comment:
Isn't this redundant? `SPIN_WAIT` already calls `cpu_relax()` on each iteration.

Reply:
Agree, redundant. Fixed in d4effb4